faucet/internal/failure/failure.go

```diff
@@ -0,0 +1,89 @@
+package failure
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+
+	"github.com/filecoin-project/faucet/internal/platform/lotus"
+)
+
+type Detector struct {
+	m                         sync.Mutex
+	log                       *logging.ZapEventLogger
+	lastBlockHeight           uint64
+	lotus                     lotus.API
+	lastBlockHeightUpdateTime time.Time
+	threshold                 time.Duration
+	checkInterval             time.Duration
+	stopChan                  chan bool
+	ticker                    *time.Ticker
+}
+
+// NewDetector creates a new failure detector that checks height value each checkInterval
+// and triggers a failure if there is no block height update in threshold.
+func NewDetector(log *logging.ZapEventLogger, api lotus.API, checkInterval, threshold time.Duration) *Detector {
+	d := Detector{
+		checkInterval:             checkInterval,
+		lotus:                     api,
+		log:                       log,
+		threshold:                 threshold,
+		stopChan:                  make(chan bool),
+		lastBlockHeightUpdateTime: time.Now(),
+		ticker:                    time.NewTicker(checkInterval),
+	}
+
+	go d.run()
+
+	return &d
+}
+
+func (d *Detector) run() {
+	for {
+		select {
+		case <-d.stopChan:
+			close(d.stopChan)
+			d.log.Infow("shutdown", "status", "detector stopped")
+			return
+		case <-d.ticker.C:
+			status, err := d.lotus.NodeStatus(context.Background(), true)
```

Comment:
When `d.StopChan` is closed, can we get stuck in the call of `d.lotus.NodeStatus`? I don't know how serious of a problem this is, feel free to ignore. A cancelable context that cancels when `d.stopChan` is closed could be a potential solution.